File: app/filter.go
package app

import "strings"

type PacketFilter struct {
	Protocol string
	SrcIp    string
	DstIp    string
}

// Функция применения фильтра сетевых пакетов
func (app *App) ApplyPacketFilter(protocol string, srcIp string, dstIp string) {
	app.Filter = PacketFilter{
		Protocol: protocol,
		SrcIp:    srcIp,
		DstIp:    dstIp,
	}
}

// Функция проверки пакета
func (app *App) CheckPacket(packet PacketInfo) bool {
	filter := app.Filter

	if filter.Protocol != "" && !strings.Contains(strings.ToLower(packet.Protocol), strings.ToLower(filter.Protocol)) {
		return false
	}

	if filter.SrcIp != "" && !strings.Contains(packet.SrcIP, filter.SrcIp) {
		return false
	}

	if filter.DstIp != "" && !strings.Contains(packet.DstIP, filter.DstIp) {
		return false
	}

	return true
}

File: src-tauri/src/network/filter.rs
use std::sync::Arc;

use once_cell::sync::Lazy;
use tokio::sync::RwLock;

use crate::network::packet::NetworkPacket;

pub static PACKET_FILTER: Lazy<Arc<RwLock<PacketFilter>>> =
  Lazy::new(|| Arc::new(RwLock::new(PacketFilter::default())));

#[derive(Clone, Default)]
pub struct PacketFilter {
  pub protocol: String,
  pub src_ip: String,
  pub dst_ip: String,
}

impl PacketFilter {
  /// Метод проверки пакета по фильтру
  pub fn check_packet(&self, packet: &NetworkPacket) -> bool {
    if !self.protocol.is_empty()
      && !packet
        .protocol
        .to_lowercase()
        .contains(&self.protocol.to_lowercase())
    {
      return false;
    }

    if !self.src_ip.is_empty() && !packet.src_ip.contains(&self.src_ip) {
      return false;
    }

    if !self.dst_ip.is_empty() && !packet.dst_ip.contains(&self.dst_ip) {
      return false;
    }

    true
  }
}
